agent/agent.go

```
@@ -0,0 +1,825 @@
+			// only guarantees a lower bound on the frequency of ticks, it does not
+			// guarantee any upper bound. This makes it impossible to reliably detect
+			// heartbeating timeouts.
+			time.Sleep(d)
+		}
+	}
+}
+
+func (h *opAgentHeartBeater) sendHeartbeat(r *hb.HeartbeatRequest) {
+	h.Lock()
+	defer h.Unlock()
+
+	logger := h.iopts.Logger()
+	_, err := h.client.Heartbeat(context.Background(), r)
+
+	// always mark the first send attempt as success to get a base time to compare against
+	if err == nil || h.lastHeartbeatTs.IsZero() {
+		h.lastHeartbeatTs = h.nowFn()
+		return
+	}
+
+	logger.Warnf("unable to send heartbeat: %v", err)
+	// check if this has been happening for past the permitted period
+	timeSinceLastSend := h.nowFn().Sub(h.lastHeartbeatTs)
+	timeout := h.agent.opts.HeartbeatTimeout()
+	if timeSinceLastSend > timeout {
+		logger.Warnf("unable to send heartbeats for %s; timing out", timeSinceLastSend.String())
+		h.timeoutCh <- struct{}{}
+	}
+}
+
+func (h *opAgentHeartBeater) stop() {
+	if h.isRunning() {
+		h.msgChan <- heartbeatMsg{stop: true}
+		h.wg.Wait()
+	}
+}
+
+func (h *opAgentHeartBeater) close() error {
+	h.stop()
+	var err error
+	if h.conn != nil {
+		err = h.conn.Close()
+		h.conn = nil
+	}
+	return err
+}
+
+func (h *opAgentHeartBeater) notifyProcessTermination(err error) {
+	h.msgChan <- heartbeatMsg{
+		processTerminate: true,
+		err:              err,
+	}
+}
+
+func (h *opAgentHeartBeater) notifyOverwrite(err error) {
+	h.msgChan <- heartbeatMsg{
+		overwritten: true,
+		err:         err,
+	}
+}
+
+type heartbeatMsg struct {
+	stop             bool
+	processTerminate bool
+	overwritten      bool
+	err              error
+}
+
+func (hm heartbeatMsg) toRPCType(req *hb.HeartbeatRequest) {
+	if hm.processTerminate {
+		req.Code = hb.HeartbeatCode_PROCESS_TERMINATION
+		req.Error = hm.err.Error()
+	} else if hm.overwritten {
+		req.Code = hb.HeartbeatCode_OVERWRITTEN
+		req.Error = hm.err.Error()
+	} else {
+		// should never happen
```

Yeah that's fair. I prefer the panic instead of propagating up the error chain because this should be an invariant. 

I'll move the code from `toRPCType` to the main heartbeat loop and added a panic instead of failing silently.